sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkOperationRetryPolicy.java

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.batch;
+
+import com.azure.cosmos.BridgeInternal;
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.TransactionalBatchOperationResult;
+import com.azure.cosmos.implementation.HttpConstants.StatusCodes;
+import com.azure.cosmos.implementation.HttpConstants.SubStatusCodes;
+import com.azure.cosmos.implementation.IRetryPolicy;
+import com.azure.cosmos.implementation.ResourceThrottleRetryPolicy;
+import com.azure.cosmos.implementation.RetryPolicyWithDiagnostics;
+import reactor.core.publisher.Mono;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+/**
+ * A container to keep retry policies and functions for bulk.
+ */
+final class BulkOperationRetryPolicy extends RetryPolicyWithDiagnostics {
+
+    private static final int MAX_RETRIES = 1;
+
+    private final ResourceThrottleRetryPolicy resourceThrottleRetryPolicy;
+    private int attemptedRetries;
+
+    BulkOperationRetryPolicy(ResourceThrottleRetryPolicy resourceThrottleRetryPolicy) {
+        this.resourceThrottleRetryPolicy = resourceThrottleRetryPolicy;
+    }
+
+    final Mono<IRetryPolicy.ShouldRetryResult> shouldRetry(final TransactionalBatchOperationResult result) {
+
+        checkNotNull(result, "expected non-null result");
+
+        // Create CosmosException for the next retry policy to understand:
+        CosmosException exception = BridgeInternal.createCosmosException(
+            result.getStatusCode(),
+            null,
+            BulkExecutorUtil.getResponseHeadersFromBatchOperationResult(result));
+
+        if (this.resourceThrottleRetryPolicy == null) {
+            return Mono.just(IRetryPolicy.ShouldRetryResult.noRetry());
+        }
+
+        return this.resourceThrottleRetryPolicy.shouldRetry(exception);
+    }
+
+    @Override
+    public Mono<IRetryPolicy.ShouldRetryResult> shouldRetry(Exception exception) {
+
+        if (this.resourceThrottleRetryPolicy == null) {
+            return Mono.just(IRetryPolicy.ShouldRetryResult.noRetry());
+        }
+
+        return this.resourceThrottleRetryPolicy.shouldRetry(exception);
+    }
+
+    boolean shouldRetryForGone(int statusCode, int subStatusCode) {
+
+        if (statusCode == StatusCodes.GONE
+            && (subStatusCode == SubStatusCodes.PARTITION_KEY_RANGE_GONE || subStatusCode == SubStatusCodes.NAME_CACHE_IS_STALE)
+            && this.attemptedRetries < MAX_RETRIES) {
```

Comment:
I have two questions here:
1. When these two exceptions happens, it will go back to the main sink and calculate the partitionKeyRangeId, but without refreshing the cache, we may get the same/old partitionKeyRangeId back. Where the cache is refreshed?

2. Currently we check for PARTITION_KEY_RANGE_GONE  and NAME_CACHE_IS_STALE exception, what about partition splitting?

Reply:
1. I can see different refreshes for different substatus codes. For pk range, I don't see any thing to handle forceRefresh. Even in function  RxPartitionKeyRangeCache::getPartitionKeyRange function, forceRefresh variable is unused. Will make 
 some change to address collection cache refresh.
2. Will add COMPLETING_SPLIT and COMPLETING_PARTITION_MIGRATION.

Thanks.

Reply:
yea, for example in GoneAndRetryWithRetryPolicy, based on different status code, we will set forcePartitionKeyRangeRefresh or forceNameCacheRefresh, which will trigger the cache refresh. 

I am a little lost here how the cache refresh will be triggered. I will read this part one more time to understand better. 

Thanks

Reply:
Thanks Rakesh.
I read the flow one more time, and I think I understand the general flow better now. 
So each bulk request will be ultimately be an RxDocumentServiceRequest, underlying GoneAndRetryWithRetryPolicy, RenameCollectionAwareClientRetryPolicy will be applied. Since we do not retry for COMPLETING_SPLIT etc for write operations, so the exceptions will be throwned and eventually captured in BulkOperationRetryPolicy. 

and InvalidPartitionExceptionRetryPolicy is doing similar collection cache refresh~